### DISCUSSION TITLE
GraphEdit Improvements

### DISCUSSION THREAD

**Discussion by Eclextic** [godotengine/godot-proposals#3999] [GUI]
As I feel that GraphEdit and GraphNode are pretty restrictive in their customizability I opened a discussion about it...

### Necessary Features:

- [ ] Allow disabling GraphNodes default ports behavior (Add a port for every child using the child's position) and add a virtual/public method to control where to put a port and how many to put (Add new type called Port and fill an array with them with position/height as a variable of said Port type => `Port().height = 10.0`)
- [ ] Allow to manually control Input

    - By either giving an option to disable Input handling and making public functions or using the existing variables to modify the parameters (Move, Zoom, Select) or
    - By having exported InputActions (File) that can be replaced with the desired Input.

### Necessary QoL:

- [ ] Have a list of neighboring connections of a GraphNode
    - (Graphnode1  <------  ConnectionGraphnode -----> GraphNode2)
    - Left: GraphNode1
    - Right: GraphNode2 
    - If possible also remove the connection list from the GraphEdit as it is extremely unnecessary and is a performance issue. (Extra memory usage)
 

- [ ] Move a lot of the processing from the GraphEdit to the GraphNode (Connection/DisconnectionRequest, DeleteNodesRequest.

### QoL:

- [ ] Add acceleration to the zoom as well as zoom mouse scroll delta.
- [ ] Line from and to icons so that you can make e.g. arrows going to the GraphNode.
- [ ]  Line gradient coloring.

‎‎

### **_Feel free to suggest more features/ideas and I'll add them to the list!_**

**Comment by Boivin92** [comment]
I'm currently making a project management tool including task dependencies and using GraphEdit is a godsent, I just wish there were a few more tools available for it.

I'd like to be able to add nodes other than GraphNodes and have them treated in a similar way _OR_ I'd like to be able to completely disable a graph node from being interacted with by the user. I thought the new "Dragable" and "Selectable" options from beta 2 would do the trick but apparently you can still click them and it pushes the nodes back to the top of the drawing stack.

Basically I want to be able to draw lines and boxes in the space of the GraphEdit, something like this:

![image](https://user-images.githubusercontent.com/8545474/194919784-54e48d3b-933c-40e9-9aa4-49c7883ecf24.png)

where the post-its would be graphNodes, with all the current possibilities, and the lines and labels would be behind them all.

**Reply by Eclextic** [reply]
That’s cool but isn’t a **Graph**Edit.
What you can do is create a scene where there is a panel and a label node. Then through script change the different StyleBoxes used as well as handling the mouse input.

Also LineRenderer is also in Godot which might help you.

### BUT if you need zoom, selection, card titles then even though GraphEdit is not intended for this use case you should probably use it

**Reply by Boivin92** [reply]
Well mostly I need the ability to link the cards together and navigate the panel, which are the two most important features of GraphEdit. But I'll figure something out :+1:
